From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andy Boxer <andyboxer@gmail.com>
Date: Sat, 4 Jul 2026 13:47:59 +0300
Subject: [PATCH] bcm47xx: flush cache for BMIPS3300 DMA sync

BMIPS3300 requires the same post-DMA cache synchronization treatment as
the other noncoherent BMIPS CPUs handled by the common DMA code.

SSB-only BCM47xx BMIPS3300 systems use 16-byte L1 cache lines. Select
MIPS_L1_CACHE_SHIFT_4 when BCMA is absent so L1_CACHE_BYTES and
ARCH_DMA_MINALIGN match the hardware.

Signed-off-by: Andy Boxer <andyboxer@gmail.com>
---
--- a/arch/mips/bcm47xx/Kconfig
+++ b/arch/mips/bcm47xx/Kconfig
@@ -4,6 +4,7 @@ if BCM47XX
 config BCM47XX_SSB
 	bool "SSB Support for Broadcom BCM47XX"
 	select SYS_HAS_CPU_BMIPS32_3300
+	select MIPS_L1_CACHE_SHIFT_4 if !BCM47XX_BCMA
 	select SSB
 	select SSB_HOST_SOC
 	select SSB_DRIVER_MIPS
--- a/arch/mips/mm/dma-noncoherent.c
+++ b/arch/mips/mm/dma-noncoherent.c
@@ -31,6 +31,7 @@ static inline bool cpu_needs_post_dma_fl
 	case CPU_R10000:
 	case CPU_R12000:
 	case CPU_BMIPS5000:
+	case CPU_BMIPS3300:
 	case CPU_LOONGSON2EF:
 	case CPU_XBURST:
 		return true;
